builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+
+        // If the cluster size is 32 we need to check where our invocation is in the subgroup, and conditionally use
+        // invocation 31 or 63's value.
+        pResult = CreateSelect(CreateICmpEQ(pClusterSize, getInt32(32)),
+            CreateSelect(CreateICmpULT(CreateMbcnt(getInt64(UINT64_MAX)), getInt32(32)), pBroadcast31, pBroadcast63),
+                pResult);
+
+        // Finish the WWM section by calling the intrinsic.
+        return CreateWWM(pResult);
+    }
+    else
+    {
+        // Start the WWM section by setting the inactive lanes.
+        Value* pResult = CreateSetInactive(pValue, CreateGroupOperationIdentity(groupOperation, pValue->getType()));
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(2)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDSSwizzle(pResult, GetDSSwizzleBitMode(1, 0, 31))), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(4)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDSSwizzle(pResult, GetDSSwizzleBitMode(2, 0, 31))), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(8)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDSSwizzle(pResult, GetDSSwizzleBitMode(4, 0, 31))), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(16)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDSSwizzle(pResult, GetDSSwizzleBitMode(8, 0, 31))), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(32)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDSSwizzle(pResult, GetDSSwizzleBitMode(16, 0, 31))), pResult);
+
+        Value* const pBroadcast31 = CreateSubgroupBroadcast(pResult, getInt32(31), instName);
+        Value* const pBroadcast63 = CreateSubgroupBroadcast(pResult, getInt32(63), instName);
+
+        // If the cluster size is 64 we always compute the value by adding together the two broadcasts.
+        pResult = CreateSelect(CreateICmpEQ(pClusterSize, getInt32(64)),
+            CreateGroupOperationArithmetic(groupOperation, pBroadcast31, pBroadcast63), pResult);
+
+        // If the cluster size is 32 we need to check where our invocation is in the subgroup, and conditionally use
+        // invocation 31 or 63's value.
+        pResult = CreateSelect(CreateICmpEQ(pClusterSize, getInt32(32)),
+            CreateSelect(CreateICmpULT(CreateMbcnt(getInt64(UINT64_MAX)), getInt32(32)), pBroadcast31, pBroadcast63),
+                pResult);
+
+        // Finish the WWM section by calling the intrinsic.
+        return CreateWWM(pResult);
+    }
+}
+
+// =====================================================================================================================
+// Create a subgroup clustered inclusive scan.
```

Comment:
per GL_KHR_shader_subgroup, cluster operation only support reduction, will it be expanded in the future?
if yes, then we needn't add pClusterSize in interface.

Reply:
We've toyed with the notion of expanding it in future, so I thought I'd future proof the interface here!